src/rewrite_infinite_loops.cpp

```diff
@@ -0,0 +1,310 @@
+/**
+ * @file rewrite_infinite_loops.cpp
+ *
+ * @author  Alex Henrie
+ * @license GPL v2+
+ */
+
+#include "rewrite_infinite_loops.h"
+
+#include "chunk.h"
+#include "uncrustify.h"
+
+
+using namespace uncrustify;
+
+
+static Chunk *find_end_brace(Chunk *pc)
+{
+   while (!chunk_is_opening_brace(pc))
+   {
+      pc = pc->GetNextNcNnl();
+   }
+   return(pc->SkipToMatch());
+}
+
+
+static bool for_needs_rewrite(Chunk *pc, E_Token desired_type)
+{
+   // The 'for' statement needs to be rewritten if `for(;;)` is not the
+   // preferred syntax for infinite loops and this 'for' is an infinite loop
+   // with no extra tokens (such as inline comments).
+
+   if (desired_type == CT_FOR)
+   {
+      return(false);
+   }
+   pc = pc->GetNext();
+
+   if (!chunk_is_token(pc, CT_SPAREN_OPEN))
+   {
+      return(false);
+   }
+   pc = pc->GetNext();
+
+   if (!chunk_is_token(pc, CT_SEMICOLON))
+   {
+      return(false);
+   }
+   pc = pc->GetNext();
+
+   if (!chunk_is_token(pc, CT_SEMICOLON))
+   {
+      return(false);
+   }
+   pc = pc->GetNext();
+
+   if (!chunk_is_token(pc, CT_SPAREN_CLOSE))
+   {
+      return(false);
+   }
+   return(true);
+}
+
+
+static bool while_needs_rewrite(Chunk *keyword, E_Token desired_type, const char *desired_condition)
+{
+   // The 'while' statement needs to be rewritten if it has only the tokens that
+   // are strictly necessary (keyword, condition, two parentheses, and semicolon
+   // if do-while) and either the keyword or the condition needs to be changed.
+
+   Chunk *oparen    = keyword->GetNext();
+   Chunk *condition = oparen->GetNext();
+   Chunk *cparen    = condition->GetNext();
+
+   if (!chunk_is_token(oparen, CT_SPAREN_OPEN))
+   {
+      return(false);
+   }
+
+   if (!chunk_is_token(cparen, CT_SPAREN_CLOSE))
+   {
+      return(false);
+   }
+
+   if (chunk_is_token(keyword, CT_WHILE_OF_DO))
+   {
+      Chunk *semicolon = cparen->GetNext();
+
+      if (!chunk_is_token(semicolon, CT_SEMICOLON))
+      {
+         return(false);
+      }
+   }
+
+   if (!chunk_is_token(keyword, desired_type))
+   {
+      return(true);
+   }
+
+   if (  strcmp(condition->Text(), "true") == 0
+      && strcmp(desired_condition, "true") != 0)
+   {
+      return(true);
+   }
+
+   if (  strcmp(condition->Text(), "1") == 0
+      && strcmp(desired_condition, "1") != 0)
+   {
+      return(true);
+   }
+   return(false);
+} // while_needs_rewrite
+
+
+static void move_one_token(Chunk **source, Chunk **destination, E_Token parent_type)
+{
+   Chunk *next_source = (*source)->GetNext();
+
+   // Place the token immediately after the previous token, without any
+   // whitespace.
+
+   (*source)->MoveAfter(*destination);
+   (*source)->column       = (*destination)->column + (*destination)->Len();
+   (*source)->orig_col     = (*destination)->orig_col + (*destination)->Len();
+   (*source)->orig_col_end = (*source)->orig_col_end + (*source)->Len();
+   (*source)->orig_prev_sp = 0;
```

Comment:
I have not done a real test, but once 'source' is moved after 'destination', do the destination's column/orig_col/orig_col_end/orig_prev_sp also need some updating? why only source's values are modified? if I am missing something obvious, please let me know.

Reply:
Still worth to check and make sure the values in the dumps are correct, even if uncrustify sorts things out fine in the end.
You can use the -ds option to see intermiediate steps and see if there is any inconsistent column values.

Reply:
I tried the `-ds` option as you suggested and it created six parse dumps, numbered `-c_002.log` through `-c_007.log`. The tokens are shown as overlapping in `-c_002.log` through `-c_004.log`. However, in `-c_005.log` through `-c_007.log`, the tokens are moved so that they do not overlap. I am not sure how to solve this, and I'm still not sure that it's a real problem. Perhaps we can leave it as it is for now, and revisit it later if it turns out to be important?

`infinite-loops.c`:

```
	do // comment
		x++;
	while (1); // comment
```

`-c_002.log`:

```
# 113>                FOR|               NONE|     PARENT_NOT_SET[  9/  9/ 12/  0][1/1/0][        c 0000][0-1]         for
# 115>        SPAREN_OPEN|                FOR|     PARENT_NOT_SET[ 12/ 12/ 14/  0][1/1/0][   2 0000 0000][0-0]            (
# 115>          SEMICOLON|                FOR|     PARENT_NOT_SET[ 13/ 13/ 16/  0][1/2/0][        8 0020][0-0]             ;
# 115>          SEMICOLON|                FOR|     PARENT_NOT_SET[ 13/ 13/ 16/  0][1/2/0][        8 0020][0-0]             ;
# 115>       SPAREN_CLOSE|                FOR|     PARENT_NOT_SET[ 14/ 14/ 17/  0][1/1/0][   2 0000 0000][0-0]              )
# 113>        VBRACE_OPEN|                 DO|     PARENT_NOT_SET[ 12/  9/  0/  0][1/1/0][             0][0-0]
# 113>        COMMENT_CPP|        COMMENT_END|     PARENT_NOT_SET[ 12/ 12/ 22/  1][2/2/0][             0][0-0]            // comment
# 113>            NEWLINE|               NONE|     PARENT_NOT_SET[ 22/ 22/ 17/  0][2/2/0][             0][1-0]
# 114>               WORD|               NONE|     PARENT_NOT_SET[ 17/ 17/ 18/  0][2/2/0][        c 0000][0-1]                 x
# 114>       INCDEC_AFTER|               NONE|     PARENT_NOT_SET[ 18/ 18/ 20/  0][2/2/0][   2 0000 0000][0-0]                  ++
# 114>          SEMICOLON|               NONE|     PARENT_NOT_SET[ 20/ 20/ 21/  0][2/2/0][   2 0000 0000][0-0]                    ;
# 114>       VBRACE_CLOSE|                 DO|     PARENT_NOT_SET[  0/ 20/  0/  0][1/1/0][             0][0-0]
# 114>            NEWLINE|               NONE|     PARENT_NOT_SET[ 21/ 21/  9/  0][1/1/0][             0][1-0]
# 115>        COMMENT_CPP|        COMMENT_END|     PARENT_NOT_SET[ 20/ 20/ 30/  1][1/1/0][             0][0-0]                    // comment
```

`-c_005.log`:

```
# 113>                FOR|               NONE|     PARENT_NOT_SET[  9/  9/ 12/  0][1/1/0][        c 0000][0-1]         for
# 115>        SPAREN_OPEN|                FOR|     PARENT_NOT_SET[ 12/ 12/ 14/  0][1/1/0][   2 0000 0000][0-0]            (
# 115>          SEMICOLON|                FOR|     PARENT_NOT_SET[ 13/ 13/ 16/  0][1/2/0][        8 0020][0-0]             ;
# 115>          SEMICOLON|                FOR|     PARENT_NOT_SET[ 14/ 13/ 16/  0][1/2/0][        8 0020][0-0]              ;
# 115>       SPAREN_CLOSE|                FOR|     PARENT_NOT_SET[ 15/ 14/ 17/  0][1/1/0][   2 0000 0000][0-0]               )
# 113>        VBRACE_OPEN|                 DO|     PARENT_NOT_SET[ 16/  9/  0/  0][1/1/0][             0][0-0]
# 113>        COMMENT_CPP|        COMMENT_END|     PARENT_NOT_SET[ 17/ 12/ 22/  1][2/2/0][             0][0-0]                 // comment
# 113>            NEWLINE|               NONE|     PARENT_NOT_SET[ 27/ 22/ 17/  0][2/2/0][             0][1-0]
# 114>               WORD|               NONE|     PARENT_NOT_SET[ 17/ 17/ 18/  0][2/2/0][        c 0000][0-1]                 x
# 114>       INCDEC_AFTER|               NONE|     PARENT_NOT_SET[ 18/ 18/ 20/  0][2/2/0][   2 0000 0000][0-0]                  ++
# 114>          SEMICOLON|               NONE|     PARENT_NOT_SET[ 20/ 20/ 21/  0][2/2/0][   2 0000 0000][0-0]                    ;
# 114>       VBRACE_CLOSE|                 DO|     PARENT_NOT_SET[ 22/ 20/  0/  0][1/1/0][             0][0-0]
# 114>            NEWLINE|               NONE|     PARENT_NOT_SET[ 22/ 21/  9/  0][1/1/0][             0][1-0]
# 115>        COMMENT_CPP|      COMMENT_WHOLE|     PARENT_NOT_SET[  9/ 20/ 30/  1][1/1/0][             0][0-0]         // comment
```